### PR TITLE
[FLINK-12834][table-planner-blink] Support CharType and BinaryType

### PR DESCRIPTION
**Pull request by JingsongLi**
## What is the purpose of the change

1.Now we use LogicalType VarcharType to support calcite char type.
2.Subsequent TableApi also generates LogicalType's CharType.
We need real support CharType in internal code gen and computation.

## Verifying this change

ut

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)

### CODE REVIEW

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/calcite/FlinkTypeFactory.scala [hunk 1]
```diff
@@ -379,14 +380,25 @@ object FlinkTypeFactory {
...
+          CharType.ofEmptyLiteral
+        } else {
+          new CharType(relDataType.getPrecision)
+        }
+      case VARCHAR =>
+        // Calcite will return a varchar RelDataType with precision 0, for example,
+        // SUBSTR('', 2, -1), Calcite infers that the return value must not exist
+        // (in this case, null will be returned when executed in Blink runner),
+        // so it infers VARCHAR(0).  But our `VarcharType` not allow precision 0.
+        new VarCharType(Math.max(1, relDataType.getPrecision))
```
Comment (JingsongLi) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/calcite/FlinkTypeFactory.scala]: Consider sql: `SELECT trim(LEADING  from '') FROM T`
`SqlTrimFunction` use `ReturnTypes.cascade(ReturnTypes.ARG2, SqlTypeTransforms.TO_NULLABLE, SqlTypeTransforms.TO_VARYING)` to infer return type.
`TO_VARYING` just convert Char to VarChar and use previous precision which maybe 0.
Now in Calcite, there is no limit to zero precision.

We need introduce an `ofEmptyLiteral()` to the variable length types as well.

Comment (JingsongLi) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/calcite/FlinkTypeFactory.scala]: After https://github.com/apache/flink/pull/8730 , it is ready now.

**Comment by JingsongLi** [comment]
This PR not resolve https://issues.apache.org/jira/browse/FLINK-10257 , Incorrect CHAR type support still remain...

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/typeutils/TypeCoercion.scala [hunk 2]
```diff
@@ -63,13 +64,14 @@ object TypeCoercion {
...
 
     case (_, DECIMAL) if isNumeric(from) => true
 
-    case (_, _) if numericWideningPrecedence.contains(from) &&
-        numericWideningPrecedence.contains(to) =>
-      if (numericWideningPrecedence.indexOf(from) < numericWideningPrecedence.indexOf(to)) {
+    case (_, _) if numericWideningPrecedence.contains(from.copy(true)) &&
+        numericWideningPrecedence.contains(to.copy(true)) =>
+      if (numericWideningPrecedence.indexOf(from.copy(true)) <
+          numericWideningPrecedence.indexOf(to.copy(true))) {
```
[the remark this answers is not in the document]
Comment (JingsongLi) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/typeutils/TypeCoercion.scala]: good idea.
just a line: `numericWideningPrecedence ++= numericWideningPrecedence.map(_.copy(false))`

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/typeutils/TypeCheckUtils.scala [hunk 3]
```diff
@@ -52,9 +52,14 @@ object TypeCheckUtils {
     case _ => false
   }
 
-  def isVarchar(dataType: LogicalType): Boolean = dataType.getTypeRoot == VARCHAR
+  def isCharacterString(dataType: LogicalType): Boolean =
+    dataType.getTypeRoot.getFamilies.contains(LogicalTypeFamily.CHARACTER_STRING)
```
[the remark this answers is not in the document]
Comment (JingsongLi) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/typeutils/TypeCheckUtils.scala]: I will just rewrite it to java and move it to runtime module.
It is not user interface, I think we should keep it in planner.

**Comment by wuchong** [comment]
LGTM

wait until travis pass

**Comment by wuchong** [comment]
@flinkbot approve all

**Pull request closed without merging**